### PR TITLE
 Adding between, and pre-ops like size & length

### PR DESCRIPTION
**Pull request by MISBMS**

### CODE REVIEW

File: src/main/java/io/appform/jsonrules/expressions/numeric/BetweenExpression.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        this.upperBound = upperBound;
+    }
+
+    @Override
+    protected boolean evaluate(ExpressionEvaluationContext context, String path, JsonNode evaluatedNode) {
+        if (null == evaluatedNode || !evaluatedNode.isNumber()) {
+            return false;
+        }
+        if (evaluatedNode.isIntegralNumber()) {
+            return evaluatedNode.asInt() > lowerBound.intValue() && evaluatedNode.asInt() < upperBound.intValue();
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/numeric/BetweenExpression.java]: Should you handle half-open-closed?

File: src/main/java/io/appform/jsonrules/expressions/preoperation/array/SizeOperation.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+public class SizeOperation extends PreOperation<Number> {
+
+    public SizeOperation() {
+        super(PreOperationType.size);
+    }
+
+    @Override
+    public Number compute(ExpressionEvaluationContext context) {
+        JsonNode node = context.getNode();
+        if (node.isArray()) {
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/preoperation/array/SizeOperation.java]: is isNull possible here?

File: src/main/java/io/appform/jsonrules/expressions/numeric/BetweenExpression.java (hunk 1 above)
Comment (MISBMS) [reply src/main/java/io/appform/jsonrules/expressions/numeric/BetweenExpression.java]: Had done for open only; moved to half closed with lower included with is the predominant case.

File: src/main/java/io/appform/jsonrules/expressions/preoperation/array/SizeOperation.java (hunk 2 above)
Comment (MISBMS) [reply src/main/java/io/appform/jsonrules/expressions/preoperation/array/SizeOperation.java]: No, if the value happens to be null, its replaced with missing node by this point.

File: src/main/java/io/appform/jsonrules/expressions/numeric/BetweenExpression.java (hunk 1 above)
Comment (santanusinha) [reply src/main/java/io/appform/jsonrules/expressions/numeric/BetweenExpression.java]: Accept option to replace <, > with <= and >= etc

**Comment by MISBMS** [comment]
too much configurability seems unnecessary in this case. if one wants
variants like >, <, >= & <= the same could be achieved through values.
Main reason for moving to half closed was, its a frequent use-case over
open.

**Comment by santanusinha** [comment]
Default can be that. Otherwise this is an incomplete implementation.

----
Upset the established order, and everything becomes chaos. I'm an agent of
chaos. Oh and you know the thing about chaos, it's fair.

File: src/main/java/io/appform/jsonrules/expressions/numeric/BetweenExpression.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    @Override
+    protected boolean evaluate(ExpressionEvaluationContext context, String path, JsonNode evaluatedNode) {
+        if (null == evaluatedNode || !evaluatedNode.isNumber()) {
+            return false;
+        }
+        boolean finalResult = false;
+        if (evaluatedNode.isIntegralNumber()) {
+            finalResult = includeLowerBound ? evaluatedNode.asInt() >= lowerBound.intValue()
+                    : evaluatedNode.asInt() > lowerBound.intValue();
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/numeric/BetweenExpression.java]: Can be long... use asLong

**Pull request merged**